x/mongo/driver/connstring/connstring.go

```diff
@@ -762,6 +771,12 @@ func (p *parser) addOption(pair string) error {
 		}
 		p.SocketTimeout = time.Duration(n) * time.Millisecond
 		p.SocketTimeoutSet = true
+	case "srvservicename":
+		// srvServiceName can only be set on URIs with the "mongodb+srv" scheme
+		if p.Scheme != SchemeMongoDBSRV {
+			return fmt.Errorf("cannot specify srvServiceName on non-SRV URI")
+		}
+		p.SRVServiceName = value
```

Comment:
This supports users configuring a `mongo.Client` with `ClientOptions#ApplyURI`. Do we need to add a `ClientOptions#SetSRVServiceName` helper? I think not. Here are my thoughts:

IIUC currently the `mongodb+srv://` scheme can only be configured through `ClientOptions#ApplyURI`. SRV lookup occurs here when parsing the URI.

A hypothetical `ClientOptions#SetSRVServiceName` would need to be applied before `ClientOptions#ApplyURI`. Example:

```
clientOpts := options.Client().ApplyURI("mongodb+srv://test1.test.build.10gen.cc").
	SetSRVServiceName("custom")
// "custom" cannot get applied. SRV lookup occurred in the call to ApplyURI.
```

So I think there's nothing to do. This may be a slight deviation from the norm. Since most (if not all) URI options have corresponding `ClientOptions#Set*` helpers. But this seems reasonable given that caveat.

Reply:
I think your investigation/explanation is correct. `ClientOptions#SetSRVServiceName` would be useless after `ApplyURI()`. Additionally, unlike fields with a `ClientOptions#Set*` helper, `SRVServiceName` is not a field on `ClientOptions` and is only a field on `ConnString`. Other `ConnString` fields like `SSL*` also have no corresponding setter on `ClientOptions`.

If we want to communicate this caveat to users, do you know of a good place to document that `srvServiceName` can only be specified in the URI and cannot be set through `ClientOptions`?

Reply:
Update: I've actually added `ClientOptions#SetSRVServiceName` and a corresponding `SRVServiceName` option. This is needed to eventually pass `SRVServiceName` to `Topology`, so we can remove duplicate parsing code. I've left a line in the documentation clarifying that if users wish to set a custom SRV service name for SRV _discovery_ with `SetSRVServiceName`, they must do so before calling `ApplyURI(...)`.

Reply:
Ah nice, that makes sense. The `SRVServiceName` option would need to be exported from the `options` package to be accessible to the `mongo` package.